### PR TITLE
Add support for refreshing specific targets.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Can be specified by the user using `pulumi refresh --target urn`.  Multiple resources can be refreshed at once.

### CODE REVIEW

File: pkg/resource/deploy/plan_executor.go [hunk 1]
```diff
@@ -303,39 +312,54 @@ func (pe *planExecutor) refresh(callerCtx context.Context, opts Options, preview
 	stepExec.SignalCompletion()
 	stepExec.WaitForCompletion()
 
-	// Rebuild this plan's map of old resources and dependency graph, stripping out any deleted resources and repairing
-	// dependency lists as necessary. Note that this updates the base snapshot _in memory_, so it is critical that any
-	// components that use the snapshot refer to the same instance and avoid reading it concurrently with this rebuild.
+	// Rebuild this plan's map of old resources and dependency graph, stripping out any deleted
+	// resources and repairing dependency lists as necessary. Note that this updates the base
+	// snapshot _in memory_, so it is critical that any components that use the snapshot refer to
+	// the same instance and avoid reading it concurrently with this rebuild.
```
Comment (CyrusNajmabadi) [inline pkg/resource/deploy/plan_executor.go]: just wrapping for this comment.

File: pkg/engine/lifecycle_test.go [hunk 2]
```diff
@@ -1554,8 +1556,37 @@ func TestRefreshWithDelete(t *testing.T) {
...
+	// refresh.
+	subsets := combinations.All(names)
+
+	// combinations.All doesn't return the empty set.  So explicitly test that case (i.e. test no
+	// targets specified)
+	validateRefreshDeleteCombination(t, names, []string{})
+
+	for _, subset := range subsets {
+		validateRefreshDeleteCombination(t, names, subset)
+	}
```
Comment (CyrusNajmabadi) [inline pkg/engine/lifecycle_test.go]: @pgavlin lmk if this is "too cute" in your testing lib.  I wanted to be able to bulk test someone trying every combinatoin of delete targets.  i.e. "A", "B", "A,B", "C", "A,C", "B,C", "A,B,C".  Was a lot easier to just write this way than duplicating tests.

File: pkg/resource/deploy/plan_executor.go [hunk 3]
```diff
@@ -303,39 +312,54 @@ func (pe *planExecutor) refresh(callerCtx context.Context, opts Options, preview
...
+			// We produces a refresh step for this specific resource.  Use the new information about
+			// its dependencies during the update.
+			old = step.Old()
+			new = step.New()
+		} else {
+			// We didn't do anything with this resource.  However, we still may want to update its
+			// dependencies.  So use this resource itself as the 'new' one to update.
+			old = s
+			new = s
+		}
```
Comment (CyrusNajmabadi) [inline pkg/resource/deploy/plan_executor.go]: this is hte part i'm the least confident about @pgavlin .  I also attempted to update the algorithm to add 'Same' steps at line 300.  However, it was unclear to me how to create a new Same-Step.  Specifically: NewSameStep needs to be passed a RegisterResourceEvent, and it wasn't clear to me how i would get such a value passed in.  Thoughts?

**Comment by CyrusNajmabadi** [comment]
@pgavlin this is ready for review.

File: pkg/engine/lifecycle_test.go (hunk 2 above)
Comment (pgavlin) [reply pkg/engine/lifecycle_test.go]: Nope, this is great

File: pkg/resource/deploy/plan_executor.go (hunk 1 above)
Comment (pgavlin) [reply pkg/resource/deploy/plan_executor.go]: Why? It's already wrapped at 120 as per the guidelines, right?

File: pkg/resource/deploy/plan_executor.go (hunk 3 above)
Comment (pgavlin) [reply pkg/resource/deploy/plan_executor.go]: > Specifically: NewSameStep needs to be passed a RegisterResourceEvent, and it wasn't clear to me how i would get such a value passed in. Thoughts?

All you need is a value that satisfies the `RegisterResourceEvent` interface. You could do something like:
```go
type registerSameEvent struct {
    goal *resource.Goal
}

func newRegisterSameEvent(state *resource.State) RegisterResourceEvent {
    goal := // create a new goal using the values in `state`
    return &registerSameEvent{goal}
}

func (e registerSameEvent) event() {}

func (e registerSameEvent) Goal() *resource.Goal {
    return e.goal
}

func (e registerSameEvent) Done(_ *RegisterResult) {}
```

Comment (pgavlin) [reply pkg/resource/deploy/plan_executor.go]: (that said, I think what you have here is fine)

Comment (CyrusNajmabadi) [reply pkg/resource/deploy/plan_executor.go]: ok.  if the existing approach is fine, def sign off and i'll merge in.  thanks!

**Pull request merged**